### PR TITLE
Initial version of the middleware chapter

### PR DESCRIPTION
**Pull request by kaloyanm**

### CODE REVIEW

File: docs/source/middleware.rst [hunk 1]
```diff
@@ -0,0 +1,61 @@
+The Middleware
+==============
+
+A middleware is a piece of code which is sitting between some kind of request 
+and the actual code that is going to be executed and ``redis-tasks`` is no 
+exception. Just before the actual running, the framework passes the task as first
+class function to the middleware. That means the worker had processed the
+information from the ``Queue`` already, gotten the task and its arguments and created a 
+middleware instance. Then it calls a method of this instance with appropriate 
+information. You are left on your own when and how to run the task.
```
Comment (vasilty) [inline docs/source/middleware.rst:10]: >You are left on your own when and how to run the task. 

This is a bit confusing

File: docs/source/middleware.rst [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+the returned result of the ``run_task`` is completely ignored. 
+
+If you want to use your custom middleware you need to include it into your 
+``rt_settings``. Let's assume that your code lies in a file named `middleware.py`
+in the root directory of your project. You have to declare it like this:
+
+.. code:: python
+
+   MIDDLEWARE = [
+      "middleware.SleepMiddleware",
```
Comment (vasilty) [inline docs/source/middleware.rst:36]: we should explain the order of middleware here

File: docs/source/middleware.rst [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+   class SleepMiddleware:
+      def run_task(self, task, run, args, kwargs):
+         time.sleep(1)
+         run(*args, *kwargs)
+
+You need to declare it as a plain python ``class`` and it must have method called 
+``run_task``. The most important parameter here is ``run``. As we mentioned above the 
+framework leaves the calling in your hands. You don't need to return anything since
+the returned result of the ``run_task`` is completely ignored.
```
Comment (dkusnik) [inline docs/source/middleware.rst:27]: The middleware definition is a simple Python class, which must have a method called ``run_task``. The function takes two obligatory parameters - `task` and `run` which define respectively the name of the task and the reference to the function which should be executed. As can be seen, the code before the execution of the `run` call is what is of our interest.

File: docs/source/middleware.rst [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+         run(*args, *kwargs)
+
+You need to declare it as a plain python ``class`` and it must have method called 
+``run_task``. The most important parameter here is ``run``. As we mentioned above the 
+framework leaves the calling in your hands. You don't need to return anything since
+the returned result of the ``run_task`` is completely ignored. 
+
+If you want to use your custom middleware you need to include it into your 
+``rt_settings``. Let's assume that your code lies in a file named `middleware.py`
+in the root directory of your project. You have to declare it like this:
```
Comment (dkusnik) [inline docs/source/middleware.rst:31]: In order to use the previously defined Middleware additional entries to the ``rt_settings.py`` file are necessary. It is a good practice to put the middleware definitions in a separate directory, e.g. `middleware`, Then the definition would be:

File: docs/source/middleware.rst [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+.. code:: python
+
+   MIDDLEWARE = [
+      "middleware.SleepMiddleware",
+   ]
+
+But wait, what if you want to do stuff right after the task is done? You use the
+same approach as above but instead of declaring ``run_task`` method you name it 
+``process_outcome``. In the example below we are sending a report via an email:
```
Comment (dkusnik) [inline docs/source/middleware.rst:41]: So the code after `run()` in the `run_task` method would not be executed? Maybe it should emphasized.

File: docs/source/middleware.rst [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+   class ReportMiddleware:
+      def process_outcome(self, task, *exc_info):
+         msg = EmailMessage()
+         msg['Subject'] = 'Daily Report'
+         msg['From'] = 'development@test.com'
+         msg['To'] = 'client@test.com'
+
+         s = smtplib.SMTP('localhost')
+         s.send_message(msg)
+         s.quit()
```
Comment (dkusnik) [inline docs/source/middleware.rst:57]: what does the `task` parameter mean?

File: docs/source/middleware.rst [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+And again you don't have the result of the task here. You should either use
+database, filesystem or even remote resource. By adding it to
+``rt_settings`` you are making it active, just like the first example.
```
Comment (i-trofimtschuk) [inline docs/source/middleware.rst:61]: reference the exact setting
add example (next iteration)

File: docs/source/middleware.rst (hunk 3 above)
Comment (i-trofimtschuk) [reply docs/source/middleware.rst:27]:
```suggestion
The middleware definition is a simple Python class, which must have a method called run_task. The function takes two obligatory parameters - task and run which define respectively the name of the task and the reference to the function which should be executed. As can be seen, the code before the execution of the run call is what is of our interest.
```

File: docs/source/middleware.rst (hunk 4 above)
Comment (i-trofimtschuk) [inline docs/source/middleware.rst:31]:
```suggestion
In order to use the previously defined Middleware additional entries to the rt_settings.py file are necessary. It is a good practice to put the middleware definitions in a separate directory, e.g. middleware, Then the definition would be:
```

File: docs/source/middleware.rst (hunk 6 above)
Comment (i-trofimtschuk) [reply docs/source/middleware.rst:57]: `task` is the task that produced the outcome

File: docs/source/middleware.rst (hunk 5 above)
Comment (i-trofimtschuk) [reply docs/source/middleware.rst:41]: process_outcome is always run, regardless of exceptions etc.

**Pull request closed without merging**